### ISSUE TITLE
[AMD][kernel][Tier 4] Port sol-execbench softmax (h=4096, f32) to FlyDSL

### ISSUE THREAD

**Issue by sunxxuns** [amdpilot-org/FlyDSL#43]
# FlyDSL implementation: port sol-execbench's softmax_h4096 problem to FlyDSL

Cross-references:
- `amdpilot-org/SOL-ExecBench` — AMD fork of `nvidia/sol-execbench` (the GPU kernel eval framework that scores AI-generated kernel solutions for correctness + performance against a PyTorch reference). Currently NVIDIA-only: CUDA / Triton / CUTLASS / CuTe DSL / cuDNN. Zero AMD support.
- `examples/cudnn/softmax/` in that fork — the softmax problem in the suite. The `definition.json` names it `softmax_h4096`; the reference is `torch.softmax(input, dim=-1)` over the last dim of a `[batch_size, 4096]` float32 tensor. The shipped solution is cuDNN-only (`cudnnSoftmaxForward` with `CUDNN_SOFTMAX_ACCURATE` + `CUDNN_SOFTMAX_MODE_INSTANCE`) and will not run on AMD.
- Sibling FlyDSL kernel issue: #38 (rmsnorm, Tier 4) is the first coverage target. This issue is the natural second target — same row-wise-reduction shape, same hidden_size=4096, but with a three-pass kernel (max → exp+sum → normalize) instead of two passes (sumsq → normalize).

## Why this issue exists

sol-execbench scores AI-generated kernel solutions across five NVIDIA-only
solution languages (PyTorch/Triton/CUTLASS/CuTe DSL/cuDNN). For the AMD port
we need FlyDSL to be a scoreable solution language, and we need coverage of
the suite's reduction-family problems beyond rmsnorm.

The `softmax_h4096` problem is the right second target because:

- It is the next-simplest kernel in the suite after rmsnorm: one row-wise
  max-reduce, one row-wise sum-reduce, with an elementwise exp+subtract+
  divide in between. No fusion, no masking, no attention, no MoE.
- The reference is plain PyTorch (`torch.softmax(input, dim=-1)`), so the
  correctness oracle is unambiguous.
- Unlike rmsnorm, the shipped solution is **cuDNN-only** — there is no
  Triton solution in `examples/triton/` for softmax. The baseline is
  therefore PyTorch eager + a Triton port written for benchmarking.
- FlyDSL already ships a production `kernels/softmax_kernel.py` (three-pass
  max/exp+sum/normalize, uses `exp2(x * log2e)` for fast exponentiation,
  register-buffers the row across passes, f32 + bf16, gfx942 + gfx950).
  The work here is **not** "write a new softmax kernel from scratch" — it
  is "wrap the existing FlyDSL softmax in the sol-execbench solution
  contract, prove parity vs the PyTorch reference, and benchmark vs a
  Triton port."
- **Known gap:** the existing `softmax_kernel.py` fast path is currently
  disabled (`if const_expr(False and N >= tile_cols ...)` at line 113),
  so only the scalar generic path runs. Re-enabling the vectorized fast
  path for `hidden_size=4096` (which is `BLOCK_THREADS * VEC_WIDTH` =
  `256 * 8` = 2048, so 4096 is a clean multiple of `tile_cols`) is in
  scope for this issue — without it the performance target is unlikely
  to be met.

## Goal (one sentence)

Implement, unit-test, and benchmark a FlyDSL softmax kernel that satisfies
the `softmax_h4096` definition from `amdpilot-org/SOL-ExecBench`, and
report a clear pass/fail vs a Triton port baseline on both MI300X (gfx942)
and MI355X (gfx950).

## Hardware / workload

```text
Hardware:  1 × MI300X (gfx942) + 1 × MI355X (gfx950)
Models:    (none — this is a kernel benchmark, not an LLM workload)
Shape:     input  [batch_size, 4096] float32
           output [batch_size, 4096] float32
           batch_size varies 1 / 32 / 1024 (per workload.jsonl)
Baseline:  sol-execbench cuDNN reference unavailable on AMD; use PyTorch
           reference (torch.softmax(input, dim=-1)) + a Triton port
           (written for this issue, reused from the Triton baseline pool)
           as baseline. Run on AMD via the unified aiter+triton+flydsl
           Docker image.
Target:    max_abs_diff < 1e-3 vs PyTorch fp32 reference on all tested
           shapes; kernel_ms < 0.9 × Triton port baseline on >=80% of
           workload shapes.
```

The workload set (`workload.jsonl`) ships 3 concrete `batch_size` values:
1, 32, 1024. The parity/bench harness should cover the small (1), mid
(32, 256), and large (1024, 4096) regimes at minimum — the shipped set
is sparse, so the test harness should add a few extra shapes (256, 4096)
to stress the vectorized fast path and the scalar generic path
respectively.

Note on tolerance: sol-execbench's default `max_atol=1e-2` (from
`docs/workload.md`) is loose for softmax in fp32 — the operator is
numerically well-conditioned after the max-subtraction trick, and the
FlyDSL kernel uses `exp2(x * log2e)` which is bit-equivalent to `exp(x)`
up to ULP. The tighter `1e-3` target reflects what a correct fp32
softmax should hit and is still well within the default
`required_matched_ratio=0.99`.

## Scope

1. `kernels/sol_execbench/softmax.py` — FlyDSL kernel implementing
   row-wise softmax over the last dim with `exp2(x - max) * log2e`
   numerics, matching the PyTorch reference. Reuse the existing
   `kernels/softmax_kernel.py::build_softmax_module` builder where
   possible — the new file is the sol-execbench-facing wrapper (entry
   point + destination-passing-style `run()`), not a reimplementation.
2. **Re-enable the vectorized fast path** in
   `kernels/softmax_kernel.py` (currently gated by
   `if const_expr(False and ...)`). For `hidden_size=4096` and
   `BLOCK_THREADS=256, VEC_WIDTH=8`, `tile_cols=2048` and
   `N % tile_cols == 0`, so the fast path should fire. If re-enabling
   exposes a correctness or compile bug, fix it or document why the
   generic path is used instead.
3. `tests/kernels/test_sol_execbench_softmax.py` — parity harness vs
   the fp32 PyTorch reference from `definition.json`. Tests multiple
   `batch_size` values: 1, 32, 256, 1024, 4096. Random inputs at
   standard normal scale, `torch.manual_seed(0)` for determinism.
   Include a large-logit case (input scaled by 100) to stress the
   max-subtraction stability.
4. `tests/kernels/bench_sol_execbench_softmax.py` — reports
   `kernel_ms` per shape + comparison vs the Triton port baseline
   (run via the unified Docker image, same shape, same hardware).
5. `assessment.md` (<=300 words) — ranking which shapes FlyDSL
   wins/ties/loses vs Triton, with a one-paragraph hypothesis for
   the loss shapes (e.g. small-batch is bandwidth-bound and Triton's
   autotune picks a better tile; large-batch is compute-bound and
   FlyDSL's vectorized fast path should win).

## Acceptance

- **Correctness:** `max_abs_diff < 1e-3` vs the PyTorch fp32 reference
  on all tested `batch_size` values (1, 32, 256, 1024, 4096), including
  the large-logit stress case. This is tighter than sol-execbench's
  default `max_atol=1e-2` but is what a correct fp32 softmax should
  achieve after the max-subtraction trick.
- **Performance:** `kernel_ms < 0.9 × Triton_port_baseline` on >=80%
  of the 3 workload shapes in `workload.jsonl` (plus any extra shapes
  added by the harness).
- **Coverage:** works on both gfx942 (MI300X) and gfx950 (MI355X).
  The existing `softmax_kernel.py` is already arch-aware via
  `get_hip_arch()`; the wrapper must exercise both paths.

## Out of scope

- **Fused softmax + masking** — that is flash-attention score
  computation (masked softmax with causal/variable-length masks) and
  is a separate, harder problem. It gets its own issue.
- **Online softmax / flash-attention** — the two-pass
  max-then-normalize algorithm here is the naive but numerically
  stable form. Flash-attention's online softmax (tracking
  running max + running sum) is out of scope.
- **Backward pass / gradient** — sol-execbench's softmax problem is
  forward-only.
- **Other sol-execbench problems** — `nemotron_rms_norm` and the rest
  of the suite are follow-ups. This issue is the **second** coverage
  target (after #38 rmsnorm), not a sweep.
- **Modifying the sol-execbench eval harness** — the wrapper conforms
  to the existing solution contract (`entry_point`,
  `destination_passing_style`, `dependencies`); no harness changes.

## Reproduction

To run the Triton port baseline locally:

```bash
# 1. Build the unified aiter+triton+flydsl Docker image (not yet built;
#    Dockerfile at /home/xiasun/unified-rocm.Dockerfile on the AMDPilot
#    fleet — ROCm 7.x base, Triton + AIter + FlyDSL installed side-by-side).
docker build -f /home/xiasun/unified-rocm.Dockerfile -t unified-rocm:latest .

# 2. Clone the AMD fork. There is no Triton softmax solution upstream
#    (only cudnn/softmax), so the Triton port is written as part of
#    this issue's benchmarking work and lives at
#    examples/triton/softmax/kernel.py in the AMD fork.
git clone https://github.com/amdpilot-org/SOL-ExecBench
cd SOL-ExecBench/examples/triton/softmax
# (driver invocation TBD — depends on the sol-execbench runner CLI,
#  which is being wired up in parallel to this issue)
```

The FlyDSL solution is benchmarked by pointing the same driver at
`kernels/sol_execbench/softmax.py` instead of `kernel.py`.

To reproduce the correctness check standalone (no sol-execbench driver
needed):

```python
import torch
from kernels.sol_execbench.softmax import launch_softmax

input = torch.randn(batch_size, 4096, dtype=torch.float32, device='cuda')
output = torch.empty_like(input)
launch_softmax(input, output, batch_size, stream=torch.cuda.Stream())
ref = torch.softmax(input, dim=-1)
assert torch.max(torch.abs(output - ref)).item() < 1e-3
```

## Definition of done

- [ ] FlyDSL kernel implemented under `kernels/sol_execbench/softmax.py`
      (wrapping `kernels/softmax_kernel.py`).
- [ ] Vectorized fast path in `kernels/softmax_kernel.py` re-enabled
      (or, if it cannot be, a comment in the file explaining why the
      generic path is used for `hidden_size=4096`).
- [ ] Parity test passes on MI300X (gfx942): `max_abs_diff < 1e-3` on
      all tested `batch_size` values (1, 32, 256, 1024, 4096) +
      large-logit stress case, deterministic on rerun.
- [ ] Parity test passes on MI355X (gfx950): same bound.
- [ ] Benchmark shows `kernel_ms < 0.9 × Triton_port_baseline` on
      >=80% of the 3 workload shapes, on at least one of MI300X /
      MI355X.
- [ ] `assessment.md` (<=300 words) submitted with per-shape
      win/tie/loss table + hypothesis for losses.
- [ ] Issue closed with a one-line summary linking to the kernel, the
      test, the benchmark, and the assessment; names the next
      sol-execbench problem to port as a follow-up issue.

## ## Leaderboard baseline to beat

These are the current best public results on NVIDIA B200 (from the
[sol-execbench leaderboard](https://research.nvidia.com/benchmarks/sol-execbench/leaderboard))
for problem `046_attention_softmax_with_softcapping_and_dropout` (kernel_id=46). The agent's
FlyDSL kernel on MI300X/MI355X should aim to beat the **Scoring Baseline**
(optimized PyTorch) and approach the **top human/AI submissions**, normalized
by cross-arch roofline.

```text
GPU:           NVIDIA B200 (leaderboard) vs AMD MI300X/MI355X (this task)
Latency unit:  ms

SOL Bound (theoretical roofline):     0.003677 ms   (SOL-Score=1.0000, speedup=18.44×)
Scoring Baseline (optimized PyTorch): N/A ms   (SOL-Score=N/A, speedup=N/A×)
Reference Implementation (plain PyTorch): 0.083701 ms   (SOL-Score=0.3154, speedup=0.61×)

Top public submissions:
  - **Rank 1** (`Hyra`): 0.007514 ms, SOL-Score=0.8867, speedup=6.49× vs reference
  - **Rank 2** (`ac4k`): 0.008057 ms, SOL-Score=0.8747
  - **Rank 3** (`Amir M. Mir | SF Tensor`): 0.007967 ms, SOL-Score=0.8741
```

**Cross-arch note:** B200 timings are not directly comparable to MI300X/MI355X
(different GPU arch, different roofline). The agent should:
1. Measure the PyTorch reference on MI300X/MI355X (same shapes, same Docker image)
2. Compute the FlyDSL kernel's speedup vs the on-AMD PyTorch reference
3. Compare that speedup ratio to the leaderboard's `avg_speedup` ratios above
4. Target: FlyDSL speedup vs PyTorch-reference-on-AMD >= 0.8 × (Rank-1 speedup vs PyTorch-reference-on-B200)

## Definition of done

- [ ] FlyDSL kernel passes correctness (max_abs_diff < 1e-2 vs PyTorch reference on all tested workload shapes)
- [ ] FlyDSL kernel_ms measured on MI300X (gfx942) AND MI355X (gfx950) across >=5 workload shapes
- [ ] FlyDSL speedup vs PyTorch-reference-on-AMD >= 0.8 × Rank-1 speedup vs PyTorch-reference-on-B200 (=6.49× if top1 exists)
- [ ] `assessment.md` (<=300 words) ranks FlyDSL vs PyTorch eager on each shape, identifies win/tie/lose regimes
- [ ] Patch modifies actual kernel source (not just bench scripts) — bench-only patches are REJECTED

## Non-goals / rejection criteria

- Bench-only patches (no FlyDSL kernel source changes) are rejected
- Patches that set expected_value = measured_value (tautological pass) are rejected
- The metric contract must compare FlyDSL vs PyTorch reference on AMD, not an absolute ms
- "Completed" without a measured speedup ratio vs the on-AMD PyTorch reference is rejected

**Comment by sunxxuns** [comment]
## AMDPilot acceptance target

**The deliverable is a FlyDSL kernel.** A PyTorch reimplementation of this
problem -- with or without `torch.compile` -- is the *baseline*, not a result.
It does not close this issue no matter how closely it matches the reference.

### Implementation contract (hard requirement)

The submitted kernel MUST be written in FlyDSL and go through the FlyDSL
compiler. Concretely, the kernel module has to actually use the DSL, e.g.:

```python
import flydsl.compiler as flyc
from flydsl.expr import arith as ea, gpu, vector as ev
```

and dispatch through `flyc.jit` / `flyc.kernel`. A file placed under
`FlyDSL/kernels/sol_execbench/` that only calls `torch.*` ops does NOT satisfy
this issue, even if it is numerically perfect and even if it is faster.

If you conclude the port is not expressible in FlyDSL today, say so explicitly
and report exactly which DSL capability is missing. That is a legitimate,
useful outcome. Silently substituting a PyTorch implementation is not.

### Baseline definition

The baseline is the sol-execbench PyTorch reference (`reference.py`) running
eager, with `torch.compile` OFF. Measure it yourself in this run, on the same
GPU and the same workload shapes you benchmark your kernel on. Do not carry a
number over from another run, another GPU, or the issue text.

### Performance target

Your FlyDSL kernel must be **at least 10% faster than that measured baseline**
-- `kernel_ms <= 0.90 x baseline_ms` -- on **at least 80% of the tested
workload shapes**.

### Correctness target

`max_abs_diff < 1e-2` versus the PyTorch reference on **all** tested shapes. A
faster kernel that fails this is a failure, not a tradeoff.

### What you must report

Report a per-shape table containing, for every tested shape: the measured
eager baseline ms, your FlyDSL kernel ms, the ratio, and win/loss. Then state
the win rate against the 80% bar. A result without the baseline column next to
the kernel column is not checkable and will not be credited.
